prime-router/src/main/kotlin/history/SubmissionHistory.kt

```diff
@@ -341,41 +341,61 @@ class DetailedSubmissionHistory(
                 // Check if destination has already been added
                 // if it is increase item counts
                 // otherwise add it to destinations
-                val index = destinations.indexOf(descendantDest)
-                if (index >= 0) {
-                    destinations[index].itemCount += descendantDest.itemCount
-                    destinations[index].itemCountBeforeQualFilter =
-                        destinations[index].itemCountBeforeQualFilter?.plus(
-                            descendantDest.itemCountBeforeQualFilter ?: 0
-                        )
-                } else {
-                    destinations += descendantDest
+                with(destinations) {
+                    val index = indexOfFirst {
+                        it.organizationId == descendantDest.organizationId && it.service == descendantDest.service
+                    }
+                    if (index >= 0) {
+                        get(index).itemCount += descendantDest.itemCount
+                        get(index).itemCountBeforeQualFilter =
+                            get(index).itemCountBeforeQualFilter?.plus(
+                                descendantDest.itemCountBeforeQualFilter ?: 0
+                            )
+                    } else {
+                        add(descendantDest)
+                    }
                 }
             }
         }
 
-        // if all the items were filtered out, no translate actions are generated (thus no destinations)
-        // to return the expected output, we create dummy destinations that are only used in enrichment
-        // these dummies are created using the receivers defined in the logs
-        if (destinations.isEmpty() && filterLogs.isNotEmpty()) {
-            val filteredReportRows = filterLogs.map { log -> log.detail.message }
-            val filteredReportItems = filterLogs.map { log ->
-                ReportStreamFilterResultForResponse(log.detail as ReportStreamFilterResult)
-            }
+        // local function to safe add elements to nullable immutable list
+        fun <T> List<T>?.add(element: T): List<T> {
+            val tempMutableList = this?.toMutableList() ?: arrayListOf()
+            tempMutableList.add(element)
+            return tempMutableList
+        }
```

Comment:
Do the lists we're using this for need to stay immutable? If not, maybe we can make them mutable and use the normal `add`. If they do need to stay immutable, it might avoid some confusion to name this `addToImmutableList` or something, to make it clear that if a `mutableList` calls `add` it'll still use the usual one. That's 100% optional, I just got kind of confused reading this earlier because I thought this function was being used for the `add` at line 355.

Reply:
They don't have to stay immutable, that's why I was wondering if it was bad practice to have mutable lists as properties. Maybe we can discuss that at standup to see what everybody thinks about that. I updated them to mutable and it does it make it simpler. But if that turns out to be problem I can update them back.

Reply:
Sounds like a plan 👍